lib/internal/test_runner/mock.js

```diff
@@ -0,0 +1,303 @@
+
+    if (typeof methodName === 'string') {
+      // This is an object method spy.
+      ObjectDefineProperty(object, methodName, descriptor);
+    } else {
+      // This is a bare function spy. There isn't much to do here but make
+      // the mock call the original function.
+      this.#implementation = original;
+    }
+  }
+
+  trackCall(call) {
+    ArrayPrototypePush(this.#calls, call);
+  }
+
+  nextImpl() {
+    const nextCall = this.#calls.length;
+    const mock = this.#mocks.get(nextCall);
+    const impl = mock ?? this.#implementation;
+
+    if (nextCall + 1 === this.#times) {
+      this.restore();
+    }
+
+    this.#mocks.delete(nextCall);
+    return impl;
+  }
+}
+
+const { nextImpl, restore, trackCall } = MockFunctionContext.prototype;
+delete MockFunctionContext.prototype.trackCall;
+delete MockFunctionContext.prototype.nextImpl;
+
+class MockTracker {
+  #mocks;
+
+  constructor() {
+    this.#mocks = [];
+  }
+
+  fn(
+    original = function() {},
+    implementation = original,
+    options = kEmptyObject
+  ) {
+    if (original !== null && typeof original === 'object') {
+      options = original;
+      original = function() {};
+      implementation = original;
+    } else if (implementation !== null && typeof implementation === 'object') {
+      options = implementation;
+      implementation = original;
+    }
+
+    validateFunction(original, 'original');
+    validateFunction(implementation, 'implementation');
+    validateObject(options, 'options');
+    const { times = Infinity } = options;
+    validateTimes(times, 'options.times');
+    const ctx = new MockFunctionContext(implementation, { original }, times);
+    return this.#setupMock(ctx, original);
+  }
+
+  method(
+    object,
+    methodName,
+    implementation = kDefaultFunction,
+    options = kEmptyObject
+  ) {
+    validateObject(object, 'object');
+    validateStringOrSymbol(methodName, 'methodName');
+
+    if (implementation !== null && typeof implementation === 'object') {
+      options = implementation;
+      implementation = kDefaultFunction;
+    }
+
+    validateFunction(implementation, 'implementation');
+    validateObject(options, 'options');
+
+    const {
+      getter = false,
+      setter = false,
+      times = Infinity,
+    } = options;
+
+    validateBoolean(getter, 'options.getter');
+    validateBoolean(setter, 'options.setter');
+    validateTimes(times, 'options.times');
+
+    if (setter && getter) {
+      throw new ERR_INVALID_ARG_VALUE(
+        'options.setter', setter, "cannot be used with 'options.getter'"
+      );
+    }
+
+    const descriptor = ObjectGetOwnPropertyDescriptor(object, methodName);
+    let original;
+
+    if (getter) {
+      original = descriptor.get;
+    } else if (setter) {
+      original = descriptor.set;
+    } else {
+      original = descriptor.value;
+    }
+
+    if (typeof original !== 'function') {
+      throw new ERR_INVALID_ARG_VALUE(
```

Comment:
What's the reasoning for using `ERR_INVALID_ARG_VALUE` instead of `ERR_INVALID_ARG_TYPE` (and `validateFunction`)?

Reply:
I thought `ERR_INVALID_ARG_VALUE` gave a nicer, more configurable, error message for this case:

> The "methodName" argument must be of type function. Received null

vs.

> The argument 'methodName' must be a method. Received null